internal/fs/buffer/inmemory_write_buffer_test.go

```diff
@@ -41,24 +46,240 @@ func (t *MemoryBufferTest) SetUp(ti *TestInfo) {
+	// Allocate a buffer
+	t.TestInitializeInMemoryBuffer()
+	data1 := generateRandomData(MiB)
+	data2 := generateRandomData(MiB)
+	data3 := generateRandomData(MiB)
+	data4 := generateRandomData(MiB)
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data3, 2*MiB)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data4, 3*MiB)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, 4*MiB)
+	AssertEq(true, bytes.Equal(data3, t.mb.flushedBuffer[0:MiB]))
+	AssertEq(true, bytes.Equal(data4, t.mb.currentBuffer[0:MiB]))
+}
+
+func (t *MemoryBufferTest) TestMultipleRandomWritesToInMemoryBuffer() {
+	// Allocate a buffer
+	t.TestInitializeInMemoryBuffer()
+
+	// Write to buffer
+	err := t.mb.WriteAt([]byte("Taco"), 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt([]byte("Burrito"), 2)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt([]byte("Pizza"), 7)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, 12)
+	AssertEq(true, bytes.Equal([]byte("TaBurriPizza"), t.mb.currentBuffer[0:t.mb.fileSize]))
+}
+
+func (t *MemoryBufferTest) TestWriteJustBeforeChunkSizeOffset() {
+	// Allocate a buffer
+	t.TestInitializeInMemoryBuffer()
+
+	// Write to buffer
+	err := t.mb.WriteAt([]byte("Taco"), MiB-1)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, MiB+3)
+	AssertEq(true, bytes.Equal([]byte("T"), t.mb.flushedBuffer[MiB-1:MiB]))
+	AssertEq(true, bytes.Equal([]byte("aco"), t.mb.currentBuffer[0:3]))
+}
+
+func (t *MemoryBufferTest) TestWriteAtChunkSizeOffset() {
+	// Allocate a buffer
+	t.TestInitializeInMemoryBuffer()
+
+	// Write to buffer
+	err := t.mb.WriteAt([]byte("Taco"), MiB)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, MiB+4)
+	AssertEq(true, bytes.Equal([]byte("Taco"), t.mb.currentBuffer[0:4]))
+}
+
+func (t *MemoryBufferTest) TestWriteJustAfterChunkSizeOffset() {
+	// Allocate a buffer
+	t.TestInitializeInMemoryBuffer()
+
+	// Write to buffer
+	err := t.mb.WriteAt([]byte("Taco"), MiB+1)
+
+	AssertNe(nil, err)
+	AssertEq(NonSequentialWriteError, err.Error())
+	AssertEq(t.mb.fileSize, 0)
+}
+
+func (t *MemoryBufferTest) TestWriteRollOverFrom2ndTo1stBlock() {
+	// Allocate a buffer
+	t.TestInitializeInMemoryBuffer()
+	data1 := generateRandomData(MiB)
+	data2 := generateRandomData(2 * KiB)
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, 2*MiB-KiB)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, 2*MiB+KiB)
+	AssertEq(true, bytes.Equal(data2[0:KiB], t.mb.flushedBuffer[MiB-KiB:MiB]))
+	AssertEq(true, bytes.Equal(data2[KiB:], t.mb.currentBuffer[0:KiB]))
+}
+
+func (t *MemoryBufferTest) TestRandomWriteOnAnAlreadyWrittenBufferBlockShouldFail() {
+	// Allocate a buffer
+	t.TestInitializeInMemoryBuffer()
+	data1 := generateRandomData(MiB)
+	data2 := generateRandomData(2 * KiB)
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
```

Comment:
Let's add test(s) for scenarios where two consecutive writes skipped more than a chunk i.e. a chunk in-between got never written to, both with cases of contentSize <= chunkSize and > chunkSize.